enhancements/network/smart-nic-ovn-offload.md

```diff
@@ -0,0 +1,389 @@
+        OvnHardwareOffload []OvnHardwareOffloadConfig `json:"ovnHardwareOffload,omitempty"`
+}
+
+type OvnHardwareOffloadConfig struct {
+	// NodeSelector matches on Labels defined in MachineConfigPoolSpec.NodeSelector
+	// OVS HWOL MachineConfigs are generated and applied to Nodes in MachineConfigPool
+	// Labels in NodeSelector are ANDed when matching on MachineConfigPoolSpec.NodeSelector
+        NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+}
+```
+
+SNO will apply SmartNIC MachineConfigs by calling Machine Config Operator APIs.
+SmartNIC MachineConfigs may include service that reverts default OVN-Kubernetes
+ovs-configuration, and dropin service that disables Open vSwitch services
+(openvswitch, ovs-vswitchd, ovsdb-server). Alternative is to move
+ovs-configuration to a MachineConfig in CNO and SNO calls CNO API to remove
+ovs-configuration service and its configuration.
+
+SNO will configure [DNS Operator placement API](https://github.com/openshift/api/blob/release-4.8/operator/v1/types_dns.go#L64) based on
+`MachineConfigPoolSelector` to isolate DNS pods from running on SmartNIC
+nodes (see [Open Questions](#open-questions) for why removing DNS pods from
+SmartNIC nodes).
+
+SNO will apply label `network.operator.openshift.io/external-ovn` (defined in
+CNO) to nodes selected by `MachineConfigPoolSelector` which drives
+ovn-controller, ovn-ipsec  and ovs-node daemonset pods away from SmartNIC
+nodes and allows scheduling of ovnkube-node-smart-nic daemonset pods onto
+SmartNIC nodes.
+
+A new config map `ovn-management-ports` will be created by SNO and mounted to
+CNO ovnkube-node-smart-nic daemonset to pass node ovn management port
+information which contains node specific SR-IOV interface names that are
+created and discovered by SNO.
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: ovn-management-ports
+  namespace: openshift-ovn-kubernetes
+data:
+  smart-nic-worker-0: '{"ovnkube-node-mgmt-port":"eth0v0"}'
+  smart-nic-worker-1: '{"ovnkube-node-mgmt-port":"eth0v0"}'
+```
+
+There are three supported modes with NVIDIA SmartNIC BlueField2: Separated
+host (default), Embedded CPU and Isolated modes. BlueField2 needs to be
+configured to Embedded CPU mode when enabling OVN offload. SR-IOV network
+config daemon, a daemonset managed by SNO, will be updated to use mstflint
+version >= `4.15.0-1.el8` which supports BlueField2 mode configuration.
+
+SNO identifies the NIC using its device ID, the NVIDIA BlueField2 device ID
+will be added support in SNO.
+
+SNO will move to use systemd service for configuring SmartNIC SR-IOV devices
+vs configuring SR-IOV devices in SR-IOV config daemonset pods. This ensures
+that SR-IOV device with VF index 0 (which is used as ovn management port)
+becomes available when ovnkube-node-smart-nic pod starts.
+
+SNO will apply udev rule to SmartNIC VF0 to ensure the VF interface name is
```

Comment:
Shall we also use MCO to apply the udev rules?

Reply:
It's possible that the pci address or interface name varies in different nodes in the same MCP, using SNO can set the node specific udev rules, while MCO may not be able to do so. we still allow user to configure different sriov policies in the same MCO pool.